File: Blind75-Practice/easy/01-two-pointers/Python/merge-strings-alternately.py
"""
LeetCode 1768: Merge Strings Alternately
Difficulty: Easy
Pattern: Two Pointers

Merge word1 and word2 by alternating letters. Append remaining letters from the longer string.

Time Complexity: O(len(word1) + len(word2))
Space Complexity: O(len(word1) + len(word2)) for the result
"""

from typing import List


class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        i, j = 0, 0
        out: List[str] = []
        while i < len(word1) and j < len(word2):
            out.append(word1[i])
            out.append(word2[j])
            i += 1
            j += 1
        if i < len(word1):
            out.append(word1[i:])
        if j < len(word2):
            out.append(word2[j:])
        return "".join(out)


if __name__ == "__main__":
    s = Solution()
    assert s.mergeAlternately("abc", "pqr") == "apbqcr"
    assert s.mergeAlternately("ab", "pqrs") == "apbqrs"
    assert s.mergeAlternately("abcd", "pq") == "apbqcd"
    print("All test cases passed!")

File: Blind75-Practice/easy/02-two-pointers/Java/MergeStringsAlternately.java
/*
LeetCode 1768: Merge Strings Alternately
Difficulty: Easy
Pattern: Two Pointers

Problem:
You are given two strings, `word1` and `word2`. Merge the strings by adding
letters in alternating order, starting with `word1`.

If one string is longer than the other, append the additional letters onto the
end of the merged string.

Example 1:
Input: word1 = "abc", word2 = "pqr"
Output: "apbqcr"
Explanation: Take one character from each string in turn.

Example 2:
Input: word1 = "ab", word2 = "pqrs"
Output: "apbqrs"
Explanation: After alternating "a/p" and "b/q", append the remaining "rs".

Example 3:
Input: word1 = "abcd", word2 = "pq"
Output: "apbqcd"
Explanation: After alternating "a/p" and "b/q", append the remaining "cd".

Time Complexity: O(n + m)
Space Complexity: O(n + m)
*/

public class MergeStringsAlternately {
    public String mergeAlternately(String word1, String word2) {


        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        // StringBuilder result = new StringBuilder();
        // int i = 0;
        // int j = 0;
        // while (i < word1.length() && j < word2.length()) {
        //     result.append(word1.charAt(i++));
        //     result.append(word2.charAt(j++));
        // }
        // if (i < word1.length()) {
        //     result.append(word1.substring(i));
        // }
        // if (j < word2.length()) {
        //     result.append(word2.substring(j));
        // }
        // return result.toString();
    }

    public static void main(String[] args) {
        MergeStringsAlternately s = new MergeStringsAlternately();
        assert s.mergeAlternately("abc", "pqr").equals("apbqcr");
        assert s.mergeAlternately("ab", "pqrs").equals("apbqrs");
        assert s.mergeAlternately("abcd", "pq").equals("apbqcd");
        System.out.println("All test cases passed!");
    }
}
